src/index.ts

```diff
@@ -28,62 +42,109 @@ export const defaultCompileOptions = (outputOptions: OutputOptions): CompileOpti
+  // the resulting code to Closure Compiler.
+  for (const transform of transforms) {
+    transform.outputOptions = outputOptions;
+    const result = await transform.preCompilation(code, 'none');
+    if (result && result.code) {
+      code = result.code;
+    }
+  }
+
+  const jsAsFile = sync(code);
+  const mapAsFile = sync('');
 
-      compileOptions = Object.assign(defaultCompileOptions(outputOptions), compileOptions, {
-        js: temp.js,
-        create_source_map: temp.map,
-      });
+  // Compile Options is the final configuration to pass into Closure Compiler.
+  // defaultCompileOptions are overrideable by ones passed in directly to the plugin
+  // but the js source and sourcemap are not overrideable, since this would break the output if passed.
+  compileOptions = {
+    ...defaultCompileOptions(transforms, outputOptions),
+    ...compileOptions,
+    ...{
+      js: jsAsFile,
+      create_source_map: mapAsFile,
+    },
+  };
 
-      const compile: Promise<string> = new Promise((resolve, reject) => {
-        new compiler(compileOptions).run((exitCode: number, stdOut: string, stdErr: string) => {
-          if (exitCode !== 0) {
-            reject(new Error(`Google Closure Compiler exit ${exitCode}: ${stdErr}`));
-          } else {
-            resolve(stdOut);
+  const compile: Promise<string> = new Promise((resolve: (stdOut: string) => void, reject: (error: any) => void) => {
+    new compiler(compileOptions).run(async (exitCode: number, code: string, stdErr: string) => {
+      if (exitCode !== 0) {
+        reject(new Error(`Google Closure Compiler exit ${exitCode}: ${stdErr}`));
+      } else {
+        // Following successful Closure Compiler compilation, each transform needs an opportunity
+        // to clean up work is performed in preCompilation via postCompilation.
+        for (const transform of transforms) {
+          const result = await transform.postCompilation(code, 'none');
+          if (result && result.code) {
+            code = result.code;
           }
-        });
-      });
+        }
+        resolve(code);
+      }
+    });
+  });
 
-      return compile.then(
-        stdOut => {
-          const sourceMap: RawSourceMap = JSON.parse(readFileSync(temp.map, 'utf8'));
-          return { code: stdOut, map: sourceMap };
-        },
-        (error: Error) => {
-          throw error;
-        },
-      );
+  return compile.then(
+    async code => {
+      return { code, map: JSON.parse(await readFile(mapAsFile, 'utf8')) };
+    },
+    (error: Error) => {
+      throw error;
+    },
+  );
+};
+
+export default function closureCompiler(compileOptions: CompileOptions = {}): Plugin {
+  let transforms: Array<Transform>;
+
+  return {
+    name: 'closure-compiler',
+    load() {
+      transforms = transforms || instantiateTransforms(this);
+    },
+    transform: async (code: string) => {
+      for (const transform of transforms) {
```

Comment:
Investigate, parallel execution to make this faster.

Reply:
Addressed with 885aace.